Check selected files from project

fuzzelhjb opened the pull request:
Add support to filter files out of a given project (cppcheck or Visual Studio project)

danmar commented on cli/cppcheckexecutor.cpp:
```diff
@@ -158,7 +158,33 @@ bool CppCheckExecutor::parseFromArgs(CppCheck *cppcheck, int argc, const char* c
+        // filter only for the selected filenames from all project files
+        std::list<ImportProject::FileSettings> newList;
+
+        for (const std::string fname : pathnames) {
```
I don't believe it's necessary to create copies of pathnames items so please make fname a reference.

danmar commented on cli/cppcheckexecutor.cpp:
```diff
@@ -158,7 +158,33 @@ bool CppCheckExecutor::parseFromArgs(CppCheck *cppcheck, int argc, const char* c
+        std::list<ImportProject::FileSettings> newList;
+
+        for (const std::string fname : pathnames) {
+            for (const ImportProject::FileSettings fsetting : settings.project.fileSettings)
```
I don't believe we need to copy fileSettings items.. so please make fsetting a reference.

fuzzelhjb commented:
ok I've updated the manual and fixed the references.

danmar commented on cli/cppcheckexecutor.cpp:
```diff
@@ -158,7 +158,33 @@ bool CppCheckExecutor::parseFromArgs(CppCheck *cppcheck, int argc, const char* c
+                    firstName = &fsetting.filename;
+                }
+
+                if (firstName->find(*secondName) != std::string::npos)
```
hmm .. this match is not very explicit. is it your intention that if `file1.cpp` is specified then `file1.cpp` in any folder will be checked? and also for instance `test-file1.cpp`?

danmar replied:
What do you think about reusing the same path matching that is used for ignoring files/paths?

danmar replied:
I am thinking about adding a utility function in lib/path.cpp for matching paths. That can be used both for ignored paths and here. What do you think?

danmar replied:
I am still a bit skeptic about this matching. If you write "1.cpp" I don't think that should match "file1.cpp".

It would feel better with something more explicit. If you want to match "file1.cpp" in any subfolder than you can write "*/file1.cpp".

If you write "src" then I think it makes sense to check the flies in "src/..." but not with the path "../foo/bar/src/.."

danmar commented:
can you describe your use case a little. Do you want to quickly be able to check only certain files? Or do you permanently want to have a filter? In the latter case I think some setting in the cppcheck-gui project file makes sense. If there are files that you permanently want to filter out then those can be excluded in the cppcheck gui project file instead..

danmar commented on man/manual.md:
```diff
@@ -136,6 +136,10 @@ You can import and use Cppcheck GUI project files in the command line tool:
 
     cppcheck --project=foobar.cppcheck
 
+In addition you can also define files that are part of the project and you want to check only these files with the given project configuration:
```
hmm. I don't remember if we implemented `-i` yet but we want to have that too here eventually.

danmar replied:
We have handling of `-i`. That should also be documented but it's not your responsibility to document it.

danmar commented:
Some kind of test would be nice also. For `-i` we have a test in `cppcheck/test/cli`.. see 8855978f8aef6a4748fa0a371ecfc2385666993c.

fuzzelhjb commented:
The idea is that I only want to check some files but with the given protect
configuration.
Some projects of mine are really huge but I don't want to analyze the whole
project if I just changed one or two files.
The files you add to the command-line don't have to be exact matches. I
think it's helpful if you want give some kind of wildcard so that you don't
have to define a big list of file names. So if you define 'test' it should
match files like foo-test for example.
Another use case of that filter is that you don't have to enter the exact
path to the files that you want to be analyzed. It's just a helper for
analyzing some files out of a big project without fiddling with a complex
setup.


Daniel Marjamäki <notifications@github.com> schrieb am Sa., 7. Dez. 2019,
08:47:

> *@danmar* commented on this pull request.
> ------------------------------
>
> In man/manual.md
> <https://github.com/danmar/cppcheck/pull/2378#discussion_r355107023>:
>
> > @@ -136,6 +136,10 @@ You can import and use Cppcheck GUI project files in the command line tool:
>
>      cppcheck --project=foobar.cppcheck
>
> +In addition you can also define files that are part of the project and you want to check only these files with the given project configuration:
>
> hmm. I don't remember if we implemented -i yet but we want to have that
> too here eventually.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/danmar/cppcheck/pull/2378?email_source=notifications&email_token=AAQTYQ6O3QIYFFT42OH4XITQXNIKTA5CNFSM4JPARKJKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCOKTDQY#pullrequestreview-328544707>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAQTYQ3HG2VDYZ5CP5JZWXTQXNIKTANCNFSM4JPARKJA>
> .
>

danmar commented:
my question was also if it was temporary or permanent. Do you want to recheck those files over and over..
If people want to check a subset of the files it might be good to add some handling for that in the project file.

fuzzelhjb commented:
No it’s just temporary. My workflow is that I’m used to check only those files out of a big project that I have changed recently and before I commit them to SVN or GIT. 

 

Von: Daniel Marjamäki <notifications@github.com> 
Gesendet: Samstag, 7. Dezember 2019 21:10
An: danmar/cppcheck <cppcheck@noreply.github.com>
Cc: fuzzelhjb <chschmidt9@gmail.com>; Author <author@noreply.github.com>
Betreff: Re: [danmar/cppcheck] Check selected files from project (#2378)

 

my question was also if it was temporary or permanent. Do you want to recheck those files over and over..
If people want to check a subset of the files it might be good to add some handling for that in the project file.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/danmar/cppcheck/pull/2378?email_source=notifications&email_token=AAQTYQ72DT6CB3SHETINAZ3QXP7IHA5CNFSM4JPARKJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGGONRY#issuecomment-562882247> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAQTYQY27G3FW6FJOYNWA3LQXP7IHANCNFSM4JPARKJA> .  <https://github.com/notifications/beacon/AAQTYQ4WXETKEOS5YXS6KBDQXP7IHA5CNFSM4JPARKJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGGONRY.gif>

fuzzelhjb commented:
> 
> 
> Some kind of test would be nice also. For `-i` we have a test in `cppcheck/test/cli`.. see [8855978](https://github.com/danmar/cppcheck/commit/8855978f8aef6a4748fa0a371ecfc2385666993c).

I wanted to add some tests for this feature but I'm not able to run the basic tests without getting 17 errors. I've just ran the test-helloworld.py in the subfolder test/cli with
pythong -m pytest test-helloworld.py 
cppcheck.exe is in the path environment.

danmar commented:
> I wanted to add some tests for this feature but I'm not able to run the basic tests without getting 17 errors. I've just ran the test-helloworld.py in the subfolder test/cli with

ok interesting. I very rarelly use Windows. but our appveyor.yml runs the tests.. it does not run `test-proj2` I wonder if that is because that fails.

maybe you can ignore the "old" errors you get. If your new test works that should be fine.

fuzzelhjb commented:
I added the new tests to test-helloworld.py. The reason why the tests failed was that in testutils.py a relative path for the cppcheck.exe was set in the else at line 53. I changed that line to call cppcheck without a relative path and then all tests run without a problem. I don't know how the tests are configured on the build server so I didn't commit the change in testutils.py.

danmar commented:
> I added the new tests to test-helloworld.py. The reason why the tests failed was that in testutils.py a relative path for the cppcheck.exe was set in the else at line 53. I changed that line to call cppcheck without a relative path and then all tests run without a problem. I don't know how the tests are configured on the build server so I didn't commit the change in testutils.py.

Interesting. Did you copy the cppcheck executable to `test/cli` before you run the script?

fuzzelhjb commented:
No I have the exe in my path environment.

Daniel Marjamäki <notifications@github.com> schrieb am Sa., 14. Dez. 2019,
09:14:

> I added the new tests to test-helloworld.py. The reason why the tests
> failed was that in testutils.py a relative path for the cppcheck.exe was
> set in the else at line 53. I changed that line to call cppcheck without a
> relative path and then all tests run without a problem. I don't know how
> the tests are configured on the build server so I didn't commit the change
> in testutils.py.
>
> Interesting. Did you copy the cppcheck executable to test/cli before you
> run the script?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/danmar/cppcheck/pull/2378?email_source=notifications&email_token=AAQTYQZXSDVBKAC2AEWYZW3QYSIUXA5CNFSM4JPARKJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEG35BAQ#issuecomment-565694594>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAQTYQYOWNEFLQRYYIHBIW3QYSIUXANCNFSM4JPARKJA>
> .
>

fuzzelhjb commented:
Oh didn't know that. The changes were necessary because I added another
file to the folder and project.

Daniel Marjamäki <notifications@github.com> schrieb am Sa., 14. Dez. 2019,
09:03:

> *@danmar* commented on this pull request.
>
> hmm.. I know I asked you to modify test-helloworld. but (stupid of me) I
> did not realize there would be so many changes in that test file. maybe
> it's better that I add a test for this later, it seems better to test this
> in "test-proj2.py" as I guess we can keep the project as is.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/danmar/cppcheck/pull/2378?email_source=notifications&email_token=AAQTYQ76SXQD4WX2I3GSNI3QYSHM5A5CNFSM4JPARKJKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCPGQATI#pullrequestreview-332202061>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAQTYQ3VJFKIHZZRVNS2TH3QYSHM5ANCNFSM4JPARKJA>
> .
>

fuzzelhjb commented:
I'll take a look on that layer and see what I can do here.

Daniel Marjamäki <notifications@github.com> schrieb am Sa., 14. Dez. 2019,
08:59:

> *@danmar* commented on this pull request.
> ------------------------------
>
> In cli/cppcheckexecutor.cpp
> <https://github.com/danmar/cppcheck/pull/2378#discussion_r357903704>:
>
> > +    if (!mSettings->project.fileSettings.empty()) {
> +        // filter only for the selected filenames from all project files
> +        std::list<ImportProject::FileSettings> newList;
> +
> +        for (const std::string &fname : pathnames) {
> +            for (const ImportProject::FileSettings &fsetting : settings.project.fileSettings)
> +            {
> +                const std::string *firstName = &fname;
> +                const std::string *secondName = &fsetting.filename;
> +                if (firstName->size() < secondName->size())
> +                {
> +                    secondName = &fname;
> +                    firstName = &fsetting.filename;
> +                }
> +
> +                if (firstName->find(*secondName) != std::string::npos)
>
> I am still a bit skeptic about this matching. If you write "1.cpp" I don't
> think that should match "file1.cpp".
>
> It would feel better with something more explicit. If you want to match
> "file1.cpp" in any subfolder than you can write "*/file1.cpp".
>
> If you write "src" then I think it makes sense to check the flies in
> "src/..." but not with the path "../foo/bar/src/.."
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/danmar/cppcheck/pull/2378?email_source=notifications&email_token=AAQTYQ75BH6FXIJSEUBG663QYSG45A5CNFSM4JPARKJKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCPGP7XY#discussion_r357903704>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAQTYQZMDVIZQJ6L4TDQKZTQYSG45ANCNFSM4JPARKJA>
> .
>

fuzzelhjb commented:
I'm not a python expert but what is wrong with this:
    prjpath = getAbsoluteProjectPath()
    ret, stdout, stderr = cppcheck(['--addon=misra', '--template=cppcheck1', prjpath])
    filename1 = os.path.join(prjpath, 'foo.c')
    filename2 = os.path.join(prjpath, 'main.c')
    assert ret == 0
    assert stderr == ('[%s:5]: (error) Memory leak: buffer\n'
                      '[%s:1]: (style) misra violation (use --rule-texts=<file> to get proper output)\n'
                      '[%s:5]: (error) Division by zero.\n'
                      '[%s:1]: (style) misra violation (use --rule-texts=<file> to get proper output)\n' % (filename1, filename1, filename2, filename2))

I always get a test error:
E       AssertionError: assert '[c:\Projects...oper output)\n' == '[c:\\Projects...per output)\n'
E         - [c:\Projects\cppcheck_fork\test\cli\helloworld\foo.c:5]: (error) Memory leak: buffer
E         + [c:\\Projects\\cppcheck_fork\\test\\cli\\helloworld\\foo.c:5]: (error) Memory leak: buffer
E         ?    +         +               +     +   +            +
E         - [c:\Projects\cppcheck_fork\test\cli\helloworld\foo.c:1]: (style) misra violation (use --rule-texts=<file> to get proper output)
E         + [c:\\Projects\\cppcheck_fork\\test\\cli\\helloworld\\foo.c:1]: (style) misra violation (use --rule-texts=<file> to get proper output)
E         ?    +         +               +     +   +            +
E         - [c:\Projects\cppcheck...
E
E         ...Full output truncated (7 lines hidden), use '-vv' to show

something is adding \\ to the path that triggers the assertion. Do you know what's wrong here?

danmar commented on cli/cppcheckexecutor.cpp:
```diff
@@ -158,7 +159,40 @@ bool CppCheckExecutor::parseFromArgs(CppCheck *cppcheck, int argc, const char* c
+            std::string searchString = fname;
+            size_t pos = fname.find('*');
+
+            // if the user added an asterix '*' replace it with regex '\w+' to match whole words
```
In suppressions, a '*' means any characters as far as I remember.

I really think that inconsistent. If you execute cppcheck like this:

    cppcheck file1.cpp

then cppcheck will not check the file `foo/bar/file1.cpp`

If you execute cppcheck like this:

    cppcheck --project=foobar.vcxproj file1.cpp

Then suddenly you want that we check the file `foo/bar/file1.cpp`.

Also imagine that the user writes:

    cppcheck --project=foobar.vcxproj a

Then as far as I see, any path that contains a "a" will be checked:

    foo/bar/file1.cpp

If I would write "a" then I would probably only want that files in the "a" folder would be checked.

Please break out this pattern matching into a function so it can be unit tested.

fuzzelhjb commented:
Ok ill change this. But the * will be replaced with \w+ that means match
any word. E.g. os_*.c would match os_api.c, os_file.c and so on.
You can even do something like os_*_app.c and will get files like
os_win_api.c or os_linux_api.c


Daniel Marjamäki <notifications@github.com> schrieb am Mo., 16. Dez. 2019,
12:32:

> *@danmar* commented on this pull request.
> ------------------------------
>
> In cli/cppcheckexecutor.cpp
> <https://github.com/danmar/cppcheck/pull/2378#discussion_r358185036>:
>
> > @@ -158,7 +159,40 @@ bool CppCheckExecutor::parseFromArgs(CppCheck *cppcheck, int argc, const char* c
>  #else
>      const bool caseSensitive = true;
>  #endif
> -    if (!pathnames.empty()) {
> +    if (!mSettings->project.fileSettings.empty()) {
> +        // filter only for the selected filenames from all project files
> +        std::list<ImportProject::FileSettings> newList;
> +
> +        for (const std::string &fname : pathnames) {
> +            std::string searchString = fname;
> +            size_t pos = fname.find('*');
> +
> +            // if the user added an asterix '*' replace it with regex '\w+' to match whole words
>
> In suppressions, a '*' means any characters as far as I remember.
>
> I really think that inconsistent. If you execute cppcheck like this:
>
> cppcheck file1.cpp
>
> then cppcheck will not check the file foo/bar/file1.cpp
>
> If you execute cppcheck like this:
>
> cppcheck --project=foobar.vcxproj file1.cpp
>
> Then suddenly you want that we check the file foo/bar/file1.cpp.
>
> Also imagine that the user writes:
>
> cppcheck --project=foobar.vcxproj a
>
> Then as far as I see, any path that contains a "a" will be checked:
>
> foo/bar/file1.cpp
>
> If I would write "a" then I would probably only want that files in the "a"
> folder would be checked.
>
> Please break out this pattern matching into a function so it can be unit
> tested.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/danmar/cppcheck/pull/2378?email_source=notifications&email_token=AAQTYQ7EAUVPNVYHAAIQJTDQY5RNZA5CNFSM4JPARKJKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCPIZ7MQ#pullrequestreview-332505010>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAQTYQYD2YUZA6J5VYOFPODQY5RNZANCNFSM4JPARKJA>
> .
>

fuzzelhjb commented:
just to be sure that I don't add the filter function at the wrong place,
where do you want the function be part of? Any utility class where this
would fit?

Christoph Schmidt <chschmidt9@gmail.com> schrieb am Mo., 16. Dez. 2019,
16:35:

> Ok ill change this. But the * will be replaced with \w+ that means match
> any word. E.g. os_*.c would match os_api.c, os_file.c and so on.
> You can even do something like os_*_app.c and will get files like
> os_win_api.c or os_linux_api.c
>
>
> Daniel Marjamäki <notifications@github.com> schrieb am Mo., 16. Dez.
> 2019, 12:32:
>
>> *@danmar* commented on this pull request.
>> ------------------------------
>>
>> In cli/cppcheckexecutor.cpp
>> <https://github.com/danmar/cppcheck/pull/2378#discussion_r358185036>:
>>
>> > @@ -158,7 +159,40 @@ bool CppCheckExecutor::parseFromArgs(CppCheck *cppcheck, int argc, const char* c
>>  #else
>>      const bool caseSensitive = true;
>>  #endif
>> -    if (!pathnames.empty()) {
>> +    if (!mSettings->project.fileSettings.empty()) {
>> +        // filter only for the selected filenames from all project files
>> +        std::list<ImportProject::FileSettings> newList;
>> +
>> +        for (const std::string &fname : pathnames) {
>> +            std::string searchString = fname;
>> +            size_t pos = fname.find('*');
>> +
>> +            // if the user added an asterix '*' replace it with regex '\w+' to match whole words
>>
>> In suppressions, a '*' means any characters as far as I remember.
>>
>> I really think that inconsistent. If you execute cppcheck like this:
>>
>> cppcheck file1.cpp
>>
>> then cppcheck will not check the file foo/bar/file1.cpp
>>
>> If you execute cppcheck like this:
>>
>> cppcheck --project=foobar.vcxproj file1.cpp
>>
>> Then suddenly you want that we check the file foo/bar/file1.cpp.
>>
>> Also imagine that the user writes:
>>
>> cppcheck --project=foobar.vcxproj a
>>
>> Then as far as I see, any path that contains a "a" will be checked:
>>
>> foo/bar/file1.cpp
>>
>> If I would write "a" then I would probably only want that files in the
>> "a" folder would be checked.
>>
>> Please break out this pattern matching into a function so it can be unit
>> tested.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/danmar/cppcheck/pull/2378?email_source=notifications&email_token=AAQTYQ7EAUVPNVYHAAIQJTDQY5RNZA5CNFSM4JPARKJKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCPIZ7MQ#pullrequestreview-332505010>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAQTYQYD2YUZA6J5VYOFPODQY5RNZANCNFSM4JPARKJA>
>> .
>>
>

danmar commented:
> where do you want the function be part of

It seems like a good place to put it in `Path`.

> But the * will be replaced with \w+ that means match any word

I am not sure we need this. If you invoke cppcheck like this:

    cppcheck --project=foobar.vcxproj *.cpp

then `*` will be handled before cppcheck is executed in the "standard" way.

danmar commented:
> Ok ill change this.

Thanks! I believe this will be a useful addition.

fuzzelhjb commented:
I updated the Path class and unit tests and fine tune the filter function a bit.

danmar commented on cli/cppcheckexecutor.cpp:
```diff
@@ -911,7 +925,7 @@ int CppCheckExecutor::check_internal(CppCheck& cppcheck, int /*argc*/, const cha
             }
         }
-        if (cppcheck.analyseWholeProgram())
+        if ( cppcheck.analyseWholeProgram())
```
nit; redundant space

danmar replied:
redundant space.

danmar commented on cli/cppcheckexecutor.cpp:
```diff
@@ -889,9 +904,8 @@ int CppCheckExecutor::check_internal(CppCheck& cppcheck, int /*argc*/, const cha
-
-            // filesettings
-            c = 0;
+        // filesettings
```
indentation of this comment seems wrong.

danmar replied:
Somehow.. the indentation for `// filesettings` is still wrong.

danmar commented on lib/path.cpp:
```diff
@@ -249,3 +250,34 @@ bool Path::fileExists(const std::string &file)
+    std::string searchString = filter;
+    size_t pos = filter.find('*');
+
+    // if the user added an asterix '*' replace it with regex '\w+' to match whole words
```
if we have globbing with `*` then it should be consistent with how other globs are working. Imho it should be constistent with how it works on the command line.

danmar commented on test/testpath.cpp:
```diff
@@ -38,6 +38,20 @@ class TestPath : public TestFixture {
+        ASSERT(Path::isMatchingFilter("c:\\foo\\bar\\os_linux.c", "os_win")==false);
+        ASSERT(Path::isMatchingFilter("c:\\foo\\bar\\os_linux.c", "bar"));
+        ASSERT(Path::isMatchingFilter("c:\\foo\\bar\\os_linux.c", "b*"));
+        ASSERT(Path::isMatchingFilter("c:\\foo\\bar\\os_linux.cpp", "*.cpp"));
```
In my opinion this should be a match because that is "de facto" behavior for globs:

    ASSERT(Path::isMatchingFilter("foo/bar/file1.cpp", "foo/bar/*.cpp"));

danmar commented on test/testpath.cpp:
```diff
@@ -38,6 +38,20 @@ class TestPath : public TestFixture {
+
+    void isMatchingFilter() const
+    {
+        ASSERT(Path::isMatchingFilter("c:\\foo\\bar\\os_linux.c", "bar"));
```
I'd recommend that in these paths we replace `\\` with `/` .. internally in cppcheck we always use the path separator `/` even in windows.

danmar commented:
Imagine that I want to only check the files that have the extension "c".. and then I might write this filter `  *.c` on the command line. For all other command line tools that would work.. i.e. `dir *.c`. But according to your rules that will match many other files like `abcd.txt`. According to your rules, if I want to check all files that ends with ".c" then I must write something like "[.]c$"?

fuzzelhjb commented:
Hmm ok understand. So how should I proceed here? Removing the filtering and
just compare filenames and patches and only on an exact match check the
files? Add a new option like --project-filter'using regex? Or restricting
the filter to just behave the same like 'dir *.cpp' ?
Do you even want such option in your tool?

Daniel Marjamäki <notifications@github.com> schrieb am Mi., 18. Dez. 2019,
10:54:

> *@danmar* commented on this pull request.
>
> I do not agree with this filtering. If you write the filenames directly on
> the command line then they should be matched according to the normal
> filename matching rules.
>
> If you really need to have a filter then a --project-filter argument
> would be better. I would suggest that we clearly state that it's a regexp
> then.
> ------------------------------
>
> In cli/cppcheckexecutor.cpp
> <https://github.com/danmar/cppcheck/pull/2378#discussion_r359238027>:
>
> > @@ -911,7 +925,7 @@ int CppCheckExecutor::check_internal(CppCheck& cppcheck, int /*argc*/, const cha
>                  c++;
>              }
>          }
> -        if (cppcheck.analyseWholeProgram())
> +        if ( cppcheck.analyseWholeProgram())
>
> nit; redundant space
> ------------------------------
>
> In cli/cppcheckexecutor.cpp
> <https://github.com/danmar/cppcheck/pull/2378#discussion_r359239051>:
>
> > @@ -889,9 +904,8 @@ int CppCheckExecutor::check_internal(CppCheck& cppcheck, int /*argc*/, const cha
>                  }
>              }
>          } else {
> -
> -            // filesettings
> -            c = 0;
> +        // filesettings
>
> indentation of this comment seems wrong.
> ------------------------------
>
> In lib/path.cpp
> <https://github.com/danmar/cppcheck/pull/2378#discussion_r359242027>:
>
> > @@ -249,3 +250,34 @@ bool Path::fileExists(const std::string &file)
>      std::ifstream f(file.c_str());
>      return f.is_open();
>  }
> +
> +bool Path::isMatchingFilter(const std::string &filename, const std::string &filter)
> +{
> +    std::string searchString = filter;
> +    size_t pos = filter.find('*');
> +
> +    // if the user added an asterix '*' replace it with regex '\w+' to match whole words
>
> if we have globbing with * then it should be consistent with how other
> globs are working. Imho it should be constistent with how it works on the
> command line.
> ------------------------------
>
> In test/testpath.cpp
> <https://github.com/danmar/cppcheck/pull/2378#discussion_r359244976>:
>
> > @@ -38,6 +38,20 @@ class TestPath : public TestFixture {
>          TEST_CASE(is_c);
>          TEST_CASE(is_cpp);
>          TEST_CASE(get_path_from_filename);
> +        TEST_CASE(isMatchingFilter);
> +    }
> +
> +    void isMatchingFilter() const
> +    {
> +        ASSERT(Path::isMatchingFilter("c:\\foo\\bar\\os_linux.c", "bar"));
> +        ASSERT(Path::isMatchingFilter("c:\\foo\\bar\\os_linux.c", "os_win")==false);
> +        ASSERT(Path::isMatchingFilter("c:\\foo\\bar\\os_linux.c", "bar"));
> +        ASSERT(Path::isMatchingFilter("c:\\foo\\bar\\os_linux.c", "b*"));
> +        ASSERT(Path::isMatchingFilter("c:\\foo\\bar\\os_linux.cpp", "*.cpp"));
>
> In my opinion this should be a match because that is "de facto" behavior
> for globs:
>
> ASSERT(Path::isMatchingFilter("foo/bar/file1.cpp", "foo/bar/*.cpp"));
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/danmar/cppcheck/pull/2378?email_source=notifications&email_token=AAQTYQ6NXDKV6RI3EZVFIU3QZHXMBA5CNFSM4JPARKJKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCPS6YOQ#pullrequestreview-333835322>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAQTYQ3BTKFM5Y6WZ5QZ3FDQZHXMBANCNFSM4JPARKJA>
> .
>

danmar commented:
> So how should I proceed here?

I think there are two acceptable approaches.

 1. Add a `--project-filter=<regexp>` option. 

 2. Reuse the standard command line globbing.. If user writes `cppcheck --project=foobar.vcxproj *.c` then the command line processor will expand that into `cppcheck --project=foobar.vcxproj file1.c file2.c file3.c` or something like that. So there will be no `*` to handle for us. If you want to check `foo/bar/file1.c` you could use the command `cppcheck --project=foobar.vcxproj */*/file1.c` at least in bash that would work directly not sure in DOS.

fuzzelhjb commented:
Ok no problem but just to get it right: if I call cppcheck --project=foobar.vcxproj *.c I won't get any file list internally at the moment. As far as I understood the code cppcheck will just parse the project file and takes all files defined there. The commandline parser just have one file and that is '*.c'.
If I just call cppcheck *.c in a directory with no .c files but with sub directories that has .c files cppcheck gives an error message: 
cppcheck: error: could not find or open any of the paths given.
It looks like cppcheck doesn't search recursively from the current directory for all files that matches the filter.

danmar commented:
Yes. I suggest that if you want to check `foo/bar/*.c` then that is the pattern you have to write. If you only write `foo` then all files (recursively) in the `foo` folder are checked. The files in `./foobar/...` should not be checked

fuzzelhjb commented:
I see that you have two different implementations for searching files recursively one for Windows and one for Linux based systems. Wouldn’t it be easier to use C++17 features to combine the searching by using std:: filesystem? It’s a C++17 feature though.

 

Von: Daniel Marjamäki <notifications@github.com> 
Gesendet: Donnerstag, 19. Dezember 2019 17:49
An: danmar/cppcheck <cppcheck@noreply.github.com>
Cc: fuzzelhjb <chschmidt9@gmail.com>; Author <author@noreply.github.com>
Betreff: Re: [danmar/cppcheck] Check selected files from project (#2378)

 

Yes. I suggest that if you want to check foo/bar/*.c then that is the pattern you have to write. If you only write foo then all files (recursively) in the foo folder are checked. The files in ./foobar/... should not be checked

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/danmar/cppcheck/pull/2378?email_source=notifications&email_token=AAQTYQ5B4ODYJD4NG3T5VMLQZOQZRA5CNFSM4JPARKJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHKG6DA#issuecomment-567570188> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAQTYQYXE3DX6LCUKD32GXDQZOQZRANCNFSM4JPARKJA> .  <https://github.com/notifications/beacon/AAQTYQ4FPSZ2OEAXHSNLG4DQZOQZRA5CNFSM4JPARKJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHKG6DA.gif>

danmar commented:
> I see that you have two different implementations for searching files recursively one for Windows and one for Linux based systems. Wouldn’t it be easier to use C++17 features to combine the searching by using std:: filesystem? It’s a C++17 feature though.

That sounds good to use someday. I want to support old compilers and systems. So maybe we'll use C++17 features in ~10 years or so. We're still not using all C++11 features.

fuzzelhjb commented:
Ok just a try. I changed the search logic a bit under Windows. If you enter something like c:\foo\bar\*.cpp but there is no cpp file in bar but in yet another subfolder it will add it to the file list. The same goes for *.c file.
What do you think?

danmar commented on cli/filelister.cpp:
```diff
@@ -43,11 +43,16 @@
+    if (PathIsDirectoryA(path.c_str()))
+        return TRUE;
 #endif
+    else  if (endsWith(path, "*.cpp", 5) || endsWith(path, "*.c", 3))
```
It looks wrong that `myIsDirectory` returns TRUE if there is a file extension.

danmar replied:
Also, `*` should be handled by the command line processor.. so how coud path end with `*.cpp`.

danmar commented on cli/filelister.cpp:
```diff
@@ -95,6 +99,17 @@ void FileLister::addFiles(std::map<std::string, std::size_t> &files, const std::
             break;
+        case 'c':   // ends with *.c
+        case 'p': { // ends with *.cpp
+            const std::string::size_type pos = searchPattern.find_last_of('\\');
```
How can there be a `\\` in the search pattern? The `Path::fromNativeSeparators` should convert given paths because internally I am expecting that the path separator is always '/'.

danmar commented:
> Ok just a try. I changed the search logic a bit under Windows. If you enter something like c:\foo\bar*.cpp but there is no cpp file in bar but in yet another subfolder it will add it to the file list. The same goes for *.c file.

I like it. However it seems to me that you are not testing it properly. How exactly do you run Cppcheck. Through the Visual Studio debugger?

If the command line says `c:\foo\bar*.cpp` then no `argv` string should contain `*`. That should be handled before cppcheck is executed by the command line processor.

fuzzelhjb commented:
Yes for testing I run Cppcheck through the VS debugger. The problem with all the fiddling of the search pattern is that under Windows FindFirstFile has problems when you search for a file pattern like *.cpp in a folder without any cpp files but with subfolders containing cpp files. For example: c:\test has no cpp files but a subfolder c:\test\foo\bar.cpp. If you enter Cppcheck c:\test\*.cpp than I would expect that Cppcheck would recursively scan all subfolders of c:\test but under Windows FindFirstFile() just stops in c:\test because no cpp file is found. That’s the reason why I extended the pattern to detect the *.cpp/*.c files.

So maybe a better approach is to always collect a list of all files in all subfolders and then filter them according to the file filter.

 

Von: Daniel Marjamäki <notifications@github.com> 
Gesendet: Mittwoch, 25. Dezember 2019 11:01
An: danmar/cppcheck <cppcheck@noreply.github.com>
Cc: fuzzelhjb <chschmidt9@gmail.com>; Author <author@noreply.github.com>
Betreff: Re: [danmar/cppcheck] Check selected files from project (#2378)

 

Ok just a try. I changed the search logic a bit under Windows. If you enter something like c:\foo\bar*.cpp but there is no cpp file in bar but in yet another subfolder it will add it to the file list. The same goes for *.c file.

I like it. However it seems to me that you are not testing it properly. How exactly do you run Cppcheck. Through the Visual Studio debugger?

If the command line says c:\foo\bar*.cpp then no argv string should contain *. That should be handled before cppcheck is executed by the command line processor.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/danmar/cppcheck/pull/2378?email_source=notifications&email_token=AAQTYQYTBIJGYLH5NEAIRWLQ2MVODA5CNFSM4JPARKJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHUGG7A#issuecomment-568877948> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAQTYQ36QKSWASW43XHVOGDQ2MVODANCNFSM4JPARKJA> .  <https://github.com/notifications/beacon/AAQTYQ4ZPKQMJYWIPS33KTDQ2MVODA5CNFSM4JPARKJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHUGG7A.gif>

danmar commented:
I don't think you understand what I'm saying.

I assume you can bypass the command processor if you run cppcheck like so:

    cppcheck "c:\test\*.cpp"

If you run cppcheck like so:

    cppcheck c:\test\*.cpp

then the command processor will be used. Your discussion about "*" does not make sense here.

Maybe the visual studio debugger does not process the * as it is done on the command line?

danmar commented:
As I said before I don't want that the files matching here:

    cppcheck "c:\test\*.cpp"

is fundamentally different to this matching:

    cppcheck --project=foo.sln "c:\test\*.cpp"

But also I am thinking it would be more consistent with standard command line usage with `--file-filter=` option or something since you want to have some custom filter behaviour.

fuzzelhjb commented:
Ok. So I add a new command line option „—file-filter=“ option and you can add a regular expression as filter. That way I don’t have to fiddle with the parser and we can use it as is. Ok?

 

Von: Daniel Marjamäki <notifications@github.com> 
Gesendet: Freitag, 27. Dezember 2019 21:56
An: danmar/cppcheck <cppcheck@noreply.github.com>
Cc: fuzzelhjb <chschmidt9@gmail.com>; Author <author@noreply.github.com>
Betreff: Re: [danmar/cppcheck] Check selected files from project (#2378)

 

As I said before I don't want that the files matching here:

cppcheck "c:\test\*.cpp"

is fundamentally different to this matching:

cppcheck --project=foo.sln "c:\test\*.cpp"

But also I am thinking it would be more consistent with standard command line usage with --file-filter= option or something since you want to have some custom filter behaviour.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/danmar/cppcheck/pull/2378?email_source=notifications&email_token=AAQTYQ6PE424PHR2SJOGIC3Q2ZTUTA5CNFSM4JPARKJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHXYBSY#issuecomment-569344203> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAQTYQ7U2EMR2M47XSFQIJLQ2ZTUTANCNFSM4JPARKJA> .  <https://github.com/notifications/beacon/AAQTYQYIHZQX72JLMB6KFR3Q2ZTUTA5CNFSM4JPARKJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHXYBSY.gif>

danmar commented:
> Ok. So I add a new command line option „—file-filter=“ option and you can add a regular expression as filter. That way I don’t have to fiddle with the parser and we can use it as is. Ok?

ok!

I would just recommend that you try out `Suppressions::matchglob` instead of regular expressions first. If that does not work as you want then you can probably use regular expressions. I don't "officially" allow regular expressions yet because that requires c++11 libraries.. but well it's probably fine.

danmar commented:
I don't understand why you need a special file pattern. What is the use case really?

If you execute Cppcheck like this:

    cppcheck c:\test

Then Cppcheck will check all files recursively in `c:\test`. So it would make sense to me that if you write:

    cppcheck --project=foobar.sln c:\test

then cppcheck will check all files in foobar.sln that are located below `c:\test`.

fuzzelhjb commented:
Ok I try to explain it again:

Let’s say you have the following project setup:

 

\project

\project\foobar\foobar.sln

\project\foobar\folder1\os_file.cpp

\project\foobar\folder1\os_api.cpp

\project\foobar\folder2\file1.cpp

 

cppcheck c:\project\foobar will check all files recursively but without the project settings in foobar.sln. 

Now let’s assume we only want to check all files starting with „os_“ but with all settings defined in foobar.sln I want to enter something like

 

cppcheck –project=\project\foobar\foobar.sln os_*.cpp

 

So Cppcheck should now filter out all project files that doesn’t meet this criteria. 

In short: I want a filter option for projects so you can just analyze some files but with the project configuration and not the complete project.

That would really help especially for huge projects.

 

 

Von: Daniel Marjamäki <notifications@github.com> 
Gesendet: Samstag, 28. Dezember 2019 07:26
An: danmar/cppcheck <cppcheck@noreply.github.com>
Cc: fuzzelhjb <chschmidt9@gmail.com>; Author <author@noreply.github.com>
Betreff: Re: [danmar/cppcheck] Check selected files from project (#2378)

 

I don't understand why you need a special file pattern. What is the use case really?

If you execute Cppcheck like this:

cppcheck c:\test

Then Cppcheck will check all files recursively in c:\test. So it would make sense to me that if you write:

cppcheck --project=foobar.sln c:\test

then cppcheck will check all files in foobar.sln that are located below c:\test.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/danmar/cppcheck/pull/2378?email_source=notifications&email_token=AAQTYQ333UWJRB4NQMIEXC3Q23WPTA5CNFSM4JPARKJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHYDNLY#issuecomment-569390767> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAQTYQZMPXUB4KDPZJNONP3Q23WPTANCNFSM4JPARKJA> .  <https://github.com/notifications/beacon/AAQTYQ4CKQGKQ7GLJ3I5VK3Q23WPTA5CNFSM4JPARKJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHYDNLY.gif>

danmar commented:
> Now let’s assume we only want to check all files starting with „os_“ but with all settings defined in foobar.sln I want to enter something like

ok.. and I guess you have `os_*.cpp` files in various deep subfolders also. thanks for the clarification. So feel free to continue with the `--file-filter`.

If everybody used Linux I'd prefer that users had to use a pipe instead:

    find -name os_\*.cpp | xargs cppcheck --project=foobar.sln

but unfortunately (for me) people don't, and there is no similar method in Windows as far as I know.

fuzzelhjb commented:
I added the file-filter option which uses `Suppressions::matchglob`. I also updated the tests and removed the old filter handling. 
However Travis seems to have a problem with something but the log output looks ok for me.

danmar commented:
> However Travis seems to have a problem with something but the log output looks ok for me.

Yes it was difficult to see. There was a warning in the selfcheck I assume that stopped Travis.

     lib/utils.h:32:22: warning: The scope of the variable 'prev' can be reduced. [variableScope]
        std::size_t cur, prev = 0;
                         ^

Maybe we should use `-q` in the self check so the warnings are not shown in the middle of the log

danmar commented on cli/cmdlineparser.cpp:
```diff
@@ -1000,6 +1004,9 @@ void CmdLineParser::printHelp()
               "                         Example: '-UDEBUG'\n"
               "    -E                   Print preprocessor output on stdout and don't do any\n"
               "                         further processing.\n"
+              "    --file-filter=<str>  Analyze only those files matching the given filter str\n"
```
Please put the options in alphabetical order.

danmar commented on cli/filelister.cpp:
```diff
@@ -77,7 +77,6 @@ void FileLister::recursiveAddFiles(std::map<std::string, std::size_t> &files, co
 void FileLister::addFiles(std::map<std::string, std::size_t> &files, const std::string &path, const std::set<std::string> &extra, bool recursive, const PathMatch& ignored)
 {
     const std::string cleanedPath = Path::toNativeSeparators(path);
-
```
this change seems redundant.

danmar commented on lib/path.cpp:
```diff
@@ -1,251 +1,252 @@
-/*
- * Cppcheck - A tool for static C/C++ code analysis
```
not sure what you do here in lib/path.cpp ... can your changes be reverted?

danmar commented on lib/settings.h:
```diff
@@ -71,6 +71,7 @@ class CPPCHECKLIB Settings : public cppcheck::Platform {
 public:
     Settings();
 
+
```
seems redundant to add this extra line.

danmar replied:
I would suggest using `git difftool master` to tweak such changes.

danmar commented on man/manual.md:
```diff
@@ -136,6 +136,16 @@ You can import and use Cppcheck GUI project files in the command line tool:
 
+In addition you can also define files that are part of the project and you want to check only these files with the given project configuration:
+    
+    cppcheck --project=foobar.cppcheck file1.cpp file2.cpp
```
This manual text needs to be updated.

fuzzelhjb commented:
I hopefully fixed all tweaks and updated the manual

danmar commented on lib/utils.h:
```diff
@@ -25,6 +25,24 @@
 #include <string>
+#include <vector>
+
+inline void splitString(const std::string &str, std::vector<std::string> &stringList, char delimeter)
```
This is not used anywhere. I suggest you remove it.

fuzzelhjb replied:
Is removed now

danmar commented:
Finally I am satisfied. If we can just make Travis and Appveyor happy we can merge this.

fuzzelhjb commented:
cool finally 💃 I merged the changes from your repro and I think we are good to go.
